File: python/tests/test_Model.py
# Copyright 2024, UChicago Argonne, LLC
# All Rights Reserved
# Software Name: NEML2 -- the New Engineering material Model Library, version 2
# By: Argonne National Laboratory
# OPEN SOURCE LICENSE (MIT)
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

import pytest
from pathlib import Path
import torch
import neml2
from neml2.tensors import Scalar, TensorType


def test_get_model():
    pwd = Path(__file__).parent
    factory = neml2.load_input(pwd / "test_Model.i")
    model = factory.get_model("model")
    assert model


def test_input_type():
    pwd = Path(__file__).parent
    model = neml2.load_model(pwd / "test_Model.i", "model")
    assert model.input_type("t") == TensorType.Scalar
    assert model.input_type("t~1") == TensorType.Scalar
    assert model.input_type("foo_rate") == TensorType.Scalar
    assert model.input_type("foo") == TensorType.Scalar
    assert model.input_type("foo~1") == TensorType.Scalar
    assert model.input_type("bar_rate") == TensorType.Scalar
    assert model.input_type("bar") == TensorType.Scalar
    assert model.input_type("bar~1") == TensorType.Scalar


def test_output_type():
    pwd = Path(__file__).parent
    model = neml2.load_model(pwd / "test_Model.i", "model")
    assert model.output_type("foo_bar_residual") == TensorType.Scalar


def test_forward():
    pwd = Path(__file__).parent
    model = neml2.load_model(pwd / "test_Model.i", "model")

    # Note input variables can have different batch shapes
    x = {
        "t": Scalar.full((1, 2), (), 0.1),
        "t~1": Scalar.zeros((3, 1, 1)),
        "foo_rate": Scalar.full(0.1),
        "foo": Scalar.full(1.5),
        "foo~1": Scalar.full((5, 2), (), 2),
        "bar_rate": Scalar.full(-0.2),
        "bar": Scalar.full(1.5),
        "bar~1": Scalar.full(2.0),
    }

    def check_y(y):
        val = y["foo_bar_residual"].torch()
        assert val.shape == (3, 5, 2)
        assert torch.allclose(val, torch.full((3, 5, 2), -0.99))

    def check_dy_dx(dy_dx):
        val = dy_dx["foo_bar_residual"]["t"].torch()
        assert val.shape == ()
        assert torch.allclose(val, torch.tensor(0.1))

        val = dy_dx["foo_bar_residual"]["t~1"].torch()
        assert val.shape == ()
        assert torch.allclose(val, torch.tensor(-0.1))

        val = dy_dx["foo_bar_residual"]["bar~1"].torch()
        assert val.shape == ()
        assert torch.allclose(val, torch.tensor(-1.0))

        val = dy_dx["foo_bar_residual"]["foo~1"].torch()
        assert val.shape == ()
        assert torch.allclose(val, torch.tensor(-1.0))

        val = dy_dx["foo_bar_residual"]["bar"].torch()
        assert val.shape == ()
        assert torch.allclose(val, torch.tensor(1.0))

        val = dy_dx["foo_bar_residual"]["bar_rate"].torch()
        assert val.shape == (3, 1, 2)
        assert torch.allclose(val, torch.tensor(-0.1))

        val = dy_dx["foo_bar_residual"]["foo"].torch()
        assert val.shape == ()
        assert torch.allclose(val, torch.tensor(1.0))

        val = dy_dx["foo_bar_residual"]["foo_rate"].torch()
        assert val.shape == (3, 1, 2)
        assert torch.allclose(val, torch.tensor(-0.1))

    y = model.value(x)
    check_y(y)

    dy_dx = model.dvalue(x)
    check_dy_dx(dy_dx)

    y, dy_dx = model.value_and_dvalue(x)
    check_y(y)
    check_dy_dx(dy_dx)

File: tests/unit/models/test_Model.cxx
// Copyright 2024, UChicago Argonne, LLC
// All Rights Reserved
// Software Name: NEML2 -- the New Engineering material Model Library, version 2
// By: Argonne National Laboratory
// OPEN SOURCE LICENSE (MIT)
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#include <catch2/catch_test_macros.hpp>
#include <catch2/matchers/catch_matchers_all.hpp>

#include "utils.h"
#include "neml2/neml2.h"
#include "neml2/models/Variable.h"
#include "neml2/tensors/tensors.h"
#include "neml2/drivers/Driver.h"

using namespace neml2;

TEST_CASE("Model", "[models]")
{
  SECTION("set_output_derivative_filter")
  {
    auto model = load_model("models/common/ScalarLinearCombination.i", "model");

    set_default_dtype(kFloat64);
    ValueMap in;
    in["A"] = Scalar::full(3.0);
    in["B"] = Scalar::full(2.0);

    SECTION("no filter returns all derivatives")
    {
      auto derivs = model->dvalue(in);
      REQUIRE(derivs.count("C") == 1);
      REQUIRE(derivs.at("C").count("A") == 1);
      REQUIRE(derivs.at("C").count("B") == 1);
    }

    SECTION("filter restricts to requested pair")
    {
      model->set_output_derivative_filter({{"C", "A"}});
      auto derivs = model->dvalue(in);
      REQUIRE(derivs.count("C") == 1);
      REQUIRE(derivs.at("C").count("A") == 1);
      REQUIRE(derivs.at("C").count("B") == 0);
    }

    SECTION("clearing filter restores all derivatives")
    {
      model->set_output_derivative_filter({{"C", "A"}});
      model->set_output_derivative_filter({});
      auto derivs = model->dvalue(in);
      REQUIRE(derivs.count("C") == 1);
      REQUIRE(derivs.at("C").count("A") == 1);
      REQUIRE(derivs.at("C").count("B") == 1);
    }

    SECTION("filter applies to value_and_dvalue")
    {
      model->set_output_derivative_filter({{"C", "B"}});
      auto [vals, derivs] = model->value_and_dvalue(in);
      REQUIRE(vals.count("C") == 1);
      REQUIRE(derivs.count("C") == 1);
      REQUIRE(derivs.at("C").count("A") == 0);
      REQUIRE(derivs.at("C").count("B") == 1);
    }

    SECTION("changing filter after evaluation uses the new filter")
    {
      // First evaluation with filter on A — traces the JIT graph
      model->set_output_derivative_filter({{"C", "A"}});
      {
        auto derivs = model->dvalue(in);
        REQUIRE(derivs.at("C").count("A") == 1);
        REQUIRE(derivs.at("C").count("B") == 0);
      }

      // Change filter to B — must invalidate the old traced graph and re-trace
      model->set_output_derivative_filter({{"C", "B"}});
      {
        auto derivs = model->dvalue(in);
        REQUIRE(derivs.at("C").count("A") == 0);
        REQUIRE(derivs.at("C").count("B") == 1);
      }
    }
  }

  SECTION("variable type")
  {
    auto model = load_model("models/common/ComposedModel3.i", "model");

    REQUIRE(model->input_variable("t").type() == TensorType::kScalar);
    REQUIRE(model->input_variable("T").type() == TensorType::kScalar);
    REQUIRE(model->input_variable("bar").type() == TensorType::kScalar);
    REQUIRE(model->input_variable("baz").type() == TensorType::kSR2);
    REQUIRE(model->input_variable("foo").type() == TensorType::kScalar);
    REQUIRE(model->output_variable("sum").type() == TensorType::kScalar);

    REQUIRE(utils::stringify(model->input_variable("t").type()) == "Scalar");
    REQUIRE(utils::stringify(model->input_variable("T").type()) == "Scalar");
    REQUIRE(utils::stringify(model->input_variable("bar").type()) == "Scalar");
    REQUIRE(utils::stringify(model->input_variable("baz").type()) == "SR2");
    REQUIRE(utils::stringify(model->input_variable("foo").type()) == "Scalar");
    REQUIRE(utils::stringify(model->output_variable("sum").type()) == "Scalar");
  }

  SECTION("graph execution exception")
  {
    auto factory = load_input("models/test_graph_execution_exception.i");
    auto driver = factory->get_driver("driver");
    REQUIRE_THROWS_WITH(
        driver->run(),
        Catch::Matchers::ContainsSubstring(
            "Try turning off just-in-time (JIT) compilation to get more detailed error messages."));
  }
}
